src/OpenApi/Schema.hs

```diff
@@ -0,0 +1,208 @@
+    Properties map  -> JSON.toJSON map
+
+newtype PropertyName = PropertyName Text
+  deriving newtype (Eq, FromJSON, FromJSONKey, Ord, ToJSON, ToJSONKey, ToText)
+  deriving stock   Show
+
+newtype MultipleOf = MultipleOf Natural
+  deriving newtype (FromJSON, ToJSON)
+  deriving stock   Show
+
+newtype ResourceId = ResourceId Text
+  deriving newtype (FromJSON, ToJSON)
+  deriving stock   Show
+
+newtype Name = Name Text
+  deriving newtype (Eq, FromJSONKey, Ord, ToJSON, ToJSONKey, ToText)
+  deriving stock   Show
+
+newtype Pattern = Pattern Text
+  deriving newtype (FromJSON, ToJSON, ToText)
+  deriving stock   Show
+
+data Schema = Reference Name | Content SchemaObject
+  deriving stock Show
+
+instance FromJSON Schema where
+  parseJSON = parseRefSum Name Content Reference "#/components/schemas/" "Schema"
+
+instance ToJSON Schema where
+  toJSON = \case
+    Content schemaObject
+      -> JSON.toJSON schemaObject
+    Reference name ->
+      JSON.object [("$ref", JSON.toJSON $ "#/components/schemas/" <> toText name)]
+
+data SchemaObject = SchemaObject
+  { additionalProperties :: Maybe AdditionalProperties
+  , allOf                :: Maybe [Schema]
+  , anyOf                :: Maybe [Schema]
+  , default'             :: Maybe JSON.Value
+  , deprecated           :: Maybe Bool
+  , description          :: Maybe Description
+  , enum                 :: Maybe Enum
+  , exclusiveMaximum     :: Maybe Bool
+  , exclusiveMinimum     :: Maybe Bool
+  , format               :: Maybe Format
+  , items                :: Maybe Schema
+  , maxLength            :: Maybe MaxLength
+  , minLength            :: Maybe MinLength
+  , not                  :: Maybe Schema
+  , nullable             :: Maybe Bool
+  , oneOf                :: Maybe [Schema]
+  , pattern'             :: Maybe Pattern
+  , properties           :: Maybe Properties
+  , required             :: Maybe [PropertyName]
+  , title                :: Maybe Title
+  , type'                :: Maybe Type
+  , uniqueItems          :: Maybe Bool
+  , xExpandableFields    :: Maybe [PropertyName]
+  , xResourceId          :: Maybe ResourceId
+  }
+  deriving stock (Generic, Show)
+
+schemaObjectRenames :: Map String String
+schemaObjectRenames = Map.fromList
+  [ ("default'",          "default")
+  , ("pattern'",          "pattern")
+  , ("type'",             "type")
+  , ("xExpandableFields", "x-expandableFields")
+  , ("xResourceId",       "x-resourceId")
+  ]
+
+instance FromJSON SchemaObject where
+  parseJSON = parseRenamed schemaObjectRenames
+
+instance ToJSON SchemaObject where
+  toJSON = generateRenamed schemaObjectRenames
+
+toJSONSchema :: SchemaObject -> JSON.Object
+toJSONSchema schemaObject = case JSON.toJSON schemaObject of
+  JSON.Object object -> collapseObject object
+  -- TODO refactor so the impure exception goes away.
+  -- It requires that we define this function natively via re-using the
+  -- generic mechanisms provided by aeson. For now it has to do.
```

Comment:
The phrase "For now it has to do." took me a bit to parse. I would probably just drop it because "it has to do" is basically -- by definition -- what all code in a final commit can be classified as anyway.